clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/producer/ProducerCompressionTest.java

```diff
@@ -0,0 +1,185 @@
+        Set<String> compressionSet = Set.of("none", "gzip", "snappy", "lz4", "zstd");
+        Set<String> protocolSet = Set.of("consumer", "classic");
+
+        compressionSet.forEach(compression -> {
+            protocolSet.forEach(protocol -> {
+                try {
+                    processCompressionTest(cluster, compression, protocol);
+                } catch (InterruptedException | ExecutionException e) {
+                    fail(e);
+                }
+            });
+        });
+    }
+
+
+    void processCompressionTest(ClusterInstance cluster, String compression, String protocol) throws InterruptedException,
+            ExecutionException {
+        String compressionTopic = topicName + "_" + protocol + "_" + compression;
+        cluster.createTopic(compressionTopic, 1, (short) 1);
+        Map<String, Object> producerProps = new HashMap<>();
+        producerProps.put(ProducerConfig.COMPRESSION_TYPE_CONFIG, compression);
+        producerProps.put(ProducerConfig.BATCH_SIZE_CONFIG, "66000");
+        producerProps.put(ProducerConfig.LINGER_MS_CONFIG, "200");
+        Producer<byte[], byte[]> producer = cluster.producer(producerProps);
+        Consumer<byte[], byte[]> consumer = cluster.consumer(Map.of(ConsumerConfig.GROUP_PROTOCOL_CONFIG, protocol));
+        try (producer) {
+            int partition = 0;
+            // prepare the messages
+            List<String> messages = IntStream.range(0, numRecords).mapToObj(this::messageValue).toList();
+            Header[] headerArr = new Header[]{new RecordHeader("key", "value".getBytes())};
+            RecordHeaders headers = new RecordHeaders(headerArr);
+
+            // make sure the returned messages are correct
+            long now = System.currentTimeMillis();
+            List<Future<RecordMetadata>> responses = new ArrayList<>();
+            messages.forEach(message -> {
+                // 1. send message without key and header
+                responses.add(producer.send(new ProducerRecord<>(compressionTopic, null, now, null,
+                        message.getBytes())));
+                // 2. send message with key, without header
+                responses.add(producer.send(new ProducerRecord<>(compressionTopic, null, now,
+                        String.valueOf(message.length()).getBytes(), message.getBytes())));
+                // 3. send message with key and header
+                responses.add(producer.send(new ProducerRecord<>(compressionTopic, null, now,
+                        String.valueOf(message.length()).getBytes(), message.getBytes(), headers)));
+            });
+            for (int offset = 0; offset < responses.size(); offset++) {
+                assertEquals(offset, responses.get(offset).get().offset(), compression);
+            }
+            verifyConsumerRecords(consumer, messages, now, headerArr, partition, compressionTopic);
+        } finally {
+            consumer.close(Duration.ofSeconds(1));
```

Comment:
Could you add some comments about why we don't use try-with-resource for consumer?